File: Arrays/Interval Scheduling without Conflicts/merge_overlapping_intervals.cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> merge(vector<vector<int>> &intervals)
{
    sort(intervals.begin(), intervals.end(), [](const std::vector<int> &a, const std::vector<int> &b)
         { return a[0] < b[0]; });
    vector<vector<int>> new_intervals;
    vector<int> singleton;
    int start = intervals[0][0];
    int end = intervals[0][1];
    int n = intervals.size();
    if (n == 1)
    {
        new_intervals.push_back(intervals[0]);
        return new_intervals;
    }
    for (int i = 1; i < n; i++)
    {
        if (intervals[i][0] <= end)
        {
            singleton.push_back(min(start, intervals[i][0]));
            singleton.push_back(max(intervals[i][1], end));
            start = min(start, intervals[i][0]);
            end = max(intervals[i][1], end);
        }
        else
        {
            if (new_intervals.empty())
            {
                new_intervals.push_back(intervals[0]);
            }
            singleton.push_back(intervals[i][0]);
            singleton.push_back(intervals[i][1]);
            start = intervals[i][0];
            end = intervals[i][1];
        }
        new_intervals.push_back(singleton);
        singleton.clear();
    }
    return new_intervals;
}

int main(){
    vector<vector<int>> intervals = {{1,4},{0,2},{3,5}};
    vector<vector<int>> result = merge(intervals);
    for (int i = 0; i < result.size(); i++) {
        cout << "[" << result[i][0] << "," << result[i][1] << "]";
    }   
    return 0;
}

File: Arrays/Interval Scheduling without Conflicts/merge_overlapping_intervals.py
def merge(intervals):
    # Sort intervals based on starting time
    intervals.sort(key=lambda x: x[0])
    new_intervals = []

    start = intervals[0][0]
    end = intervals[0][1]
    n = len(intervals)

    # If only one interval, return it directly
    if n == 1:
        new_intervals.append(intervals[0])
        return new_intervals

    for i in range(1, n):
        # If intervals overlap
        if intervals[i][0] <= end:
            start = min(start, intervals[i][0])
            end = max(end, intervals[i][1])
        else:
            # No overlap, push the current merged interval
            new_intervals.append([start, end])
            start = intervals[i][0]
            end = intervals[i][1]

    # Add the last interval
    new_intervals.append([start, end])
    return new_intervals


# Driver code
if __name__ == "__main__":
    intervals = [[1, 4], [0, 2], [3, 5]]
    result = merge(intervals)
    for interval in result:
        print(f"[{interval[0]},{interval[1]}]", end="")
